### ISSUE TITLE
[BUG] Activated state events received for background windows when clicking title bars

### ISSUE THREAD

**Issue by RedBearAK** [pop-os/cosmic-comp#892]
I just got finished adapting a Python script that was working with the `zcosmic_toplevel_info_v1` interface to use the newly updated version 2 (of the same interface). The purpose of the script is to show debugging info about the current list of running toplevel application windows, and to print out the app class and title of the "active" window that has the keyboard focus. 

Getting the active window seemed pretty simple and reliable with the version 1 `cosmic_toplevel_info_unstable_v1` protocol. But now with version 2 of the protocol and the current state of the compositor, I'm seeing something pretty odd happen with the `activated` state events. 

If I just `Alt+Tab` between windows, the debugging output shows that the active window class and title are picked up as expected. However, when I click on the window's title bar, the info in the terminal in the background shows that there was a `state` event that caused the script to set the window behind that window as the "activated" window. 

Sometimes, I have to start dragging the window before this happens. 

The window that gets seen as "active" by the script is always the one right behind the window that actually has the focus. 

Clicking elsewhere on the window, other than the title bar, shows the correct window being seen as "active". But as soon as I click on the title bar or try to drag the window, the the "active" window becomes the background window again. The only other way to fix it is to do `Alt+Tab` task switching to another window and back. 

I don't think my script is doing anything wrong, at this point. It's already filtering out some strange "empty" state events, and state events that are just arrays of four zero (x00) values. So I'm pretty sure I'm looking at some sort of bug. 

The development form of the script to query the Wayland interface is here, if anyone is interested. Other than this odd "activated" state problem, it's working as intended. 

https://github.com/RedBearAK/toshy/blob/dev_beta/wlroots-dev/query_cosmic_topl_mgmt_protocol.py

**Comment by RedBearAK** [comment]
Another thing that fixes the problem of the wrong window being seen as "active" is interacting with the maximize button. Either maximizing or unmaximizing will fix the "active" app info. Unlike clicking on the rest of the title bar, and dragging the window. 

However I've noticed that a "zero values" event happens independently of other events, if the focus is moving from a maximized window to a non-maximized window, or a maximized window is being restored to an unmaximized state. These events are confusing since a zero value would normally be interpreted as an event saying the window is maximized. But when you maximize the window you instead get a combined "maximized/activated" event. 

So I decided to filter out the lone "zero values" event, just like the empty bytes value events.

**Comment by git-f0x** [comment]
Seems possibly related to #453.

**Comment by RedBearAK** [comment]
@git-f0x 

Indeed, it does. And although there are two different claimed fixes, it is definitely not fixed yet. Thanks for the link.

**Comment by RedBearAK** [comment]
This issue is still present after getting the alpha 5 packages on a Fedora 41 test VM.

**Comment by RedBearAK** [comment]
Seems that something in PR #1772 has fixed this issue, stopping the dragged window from essentially "falling out of the focus stack" in a type of off-by-one error. Claude says it's this part, and that appears to make sense:

```rs
// Move grab is treated as focused, so don't change focus to a
// window while grab exists.
let move_grab_state = seat
    .user_data()
    .get::<SeatMoveGrabState>()
    .unwrap()
    .lock()
    .unwrap();
let move_mapped = if let Some(move_grab_state) = &*move_grab_state {
    Some(move_grab_state.element())
} else {
    None
};

let mapped = || {
    self.floating_layer
        .mapped()
        .chain(self.tiling_layer.mapped().map(|(w, _)| w))
        .chain(move_mapped.iter())  // <-- NEW: Include dragged window!
};
```

My Python test script that monitors focused window events confirms the bug I described in this issue doesn't appear to happen anymore, at least on the latest up-to-date Pop 24.04 beta. Still happening on Fedora 43 GNOME with COSMIC installed via DNF group afterward, but that should just be a matter of the Fedora packages not following the COSMIC git state as closely. I expect when they update their packages within the next few weeks to the state of the COSMIC code that includes this merged PR, this issue should go away there as well. 

Since the fix is implemented and merged into the master branch already, I suppose it makes sense to close this as resolved. 

Thanks @ids1024 for finally getting to the source of this. 🎉
